### PR TITLE
tests for output module

### PR DESCRIPTION
**Pull request by epassaro**
:warning: Requires merging https://github.com/tardis-sn/carsus-refdata/pull/3

## Motivation
`carsus` new output module has ~1000 lines and it's by far the most critical part of the code but lacks testing.

## Changes
- [x] Several changes to Azure CI pipeline to test on `macOS`
- [x] Better handling of `fname`  parameter for GFALL and Knox Long´s zeta
- [x] Replace `print` with `logging` module
- [x] Unit tests for `CMFGENReader`
- [x] Pseudo-integration tests with `pytest-notebook`:
  - [x] Check consistency between the `new` and `old` output module (SQL) by generating a small Kurucz `H-Be` + Chianti `H-He` database.
  - [x] Generate a Kurucz `H-Si` + Chianti `H-He` + CMFGEN `Si I-II` database and check against HDF5 reference data. Requires merging https://github.com/tardis-sn/carsus-refdata/pull/3

## Description
The output module takes readers and prepares the atomic data to be dumped into the HDF5 file. But several computations are performed before reaching the final DataFrames (specially with `levels` and `lines`). This computations were ported from the old SQL output module. 

I propose using the plugin `pytest-notebook` to test this module, since it requires less time to develop (and `carsus` has less manpower than `tardis`). This plugin is called with the `pytest` command plus some flags. The `.ipynb` files starting with `test_` are executed and compared with the saved output.

I wrote two notebooks. The first one compares the output of the old SQL output with the new module, including the tables preceding the final DataFrames.  This is done by creating two small atomic files (Kurucz `H-Be`, with Chianti `H-He`) during runtime. The SQL code of `carsus` is going to be removed soon, but still is useful to debug things. Once we are sure that `carsus` is mature enough to deprecate this code we should use it to avoid breaking this package.

The second notebook creates a medium size atomic file with `H-Si` Kurucz, `H-He` Chianti and `Si_I-II` CMFGEN. Then it's compared with previously stored data in `carsus-refdata` repository. This reference data is not just the final atomic file, it also includes the preceding DataFrames.

### CODE REVIEW

File: carsus/io/tests/test_cmfgen.py [hunk 1]
```diff
@@ -54,6 +74,29 @@ def si2_pho_fname(refdata_path):
...
+
+@with_refdata
+@pytest.fixture()
+def si1_reader(si1_data_dict):
+    return CMFGENReader(si1_data_dict)
+
+@with_refdata
+@pytest.fixture()
+def si2_levels_head_df():
+    return pd.read_csv(StringIO(si2_levels_head), delim_whitespace=True, names=['energy', 'j', 'label', 'method', 'priority'])
```
Comment (chvogl) [inline carsus/io/tests/test_cmfgen.py:92]: I would reformat this.

File: carsus/io/tests/test_cmfgen.py [hunk 2]
```diff
@@ -54,6 +74,29 @@ def si2_pho_fname(refdata_path):
...
+
+@with_refdata
+@pytest.fixture()
+def si2_lines_head_df():
+    return pd.read_csv(StringIO(si2_lines_head),delim_whitespace=True, names=['energy_lower', 'energy_upper', 'gf', 'j_lower',
```
Comment (chvogl) [inline carsus/io/tests/test_cmfgen.py:97]: I would reformat this.

**Review by chvogl — APPROVED** [review]
This looks good! We really needed tests for the output module. I am a little bit skeptical about the pytest-notebook but for now it is probably the best (i.e. least labor-intensive) option.

Why are the tests failing? Is it because the [refdata PR](https://github.com/tardis-sn/carsus-refdata/pull/3) has to be merged first?

There are some minor PEP8 violations. (I only marked up two). At some point, we should think about using black in Carsus.

File: carsus/io/tests/test_cmfgen.py (hunk 1 above)
Comment (epassaro) [reply carsus/io/tests/test_cmfgen.py:92]: - Yes, requires reference data merged.
- Yes, we talked about using `black` in Carsus in the near future.

**Review by andrewfullard — APPROVED** [review]

**Review by afloers — APPROVED** [review]

**Pull request merged**